src/common/transformations/src/transformations/common_optimizations/fuse_rotary_positional_embeddings.cpp

```diff
@@ -436,25 +440,53 @@ ov::pass::RoPEFusionChatGLM::RoPEFusionChatGLM(int split_output_id) {
     qkv_proj->set_output_size(3);
 
-    // get key [L, B, Hkv, S]
     auto cur_key = makePattern<opset1::Reshape>({qkv_proj->output(split_output_id), {0, 0, head_cnt, head_size}},
                                                 {{"special_zero", true}});
 
-    auto slice_Slice_437 = GenSlice(cur_key, 0, ndims, 1, 3);
-    auto var_split_1 = makePattern<opset1::VariadicSplit>({cur_key, 3, {ndims, ov::gen_pattern::Symbol("end")}});
+    std::shared_ptr<ov::Node> input_key = nullptr;
+    // Extended the RoPE to a two-dimensional form to accommodate the 2D positional encoding in GLM.
+    // Calculate positional embedding independent of batch and each head
+    if (support_2d_rope) {
+        // Get transposed key [batch, head_cnt, seq_length, head_size]
+        input_key = makePattern<opset1::Transpose>({cur_key, {0, 2, 1, 3}});
+    } else {
+        // Get key [seq_length, batch, head_cnt, head_size]
+        input_key = std::move(cur_key);
+    }
+
+    auto slice_Slice_437 = GenSlice(input_key, 0, ndims, 1, 3);
+    auto var_split_1 = makePattern<opset1::VariadicSplit>({input_key, 3, {ndims, ov::gen_pattern::Symbol("end")}});
     var_split_1->set_output_size(2);
 
     // rotate half
-    auto ListConstruct_452_Concat =
-        makePattern<opset1::Concat>({seq_length, {-1}, {head_cnt}, {ndims / 2}, {2}}, {{"axis", 0}});
-    auto const_target_shape_1 = makeConst({seq_len, batch, head_cnt, ndims / 2, 2});
-
-    auto ListConstruct_379_Concat =
-        makePattern<opset1::Concat>({seq_length, {-1}, {1}, {ndims / 2}, {2}}, {{"axis", 0}});
-    auto const_target_shape_2 = makeConst({seq_len, batch, 1, ndims / 2, 2});
+    std::shared_ptr<ov::Node> ListConstruct_452_Concat = nullptr;
+    std::shared_ptr<ov::Node> const_target_shape_1 = nullptr;
+
+    if (support_2d_rope) {
+        ListConstruct_452_Concat =
+            makePattern<opset1::Concat>({{-1}, {head_cnt}, seq_length, {ndims / 2}, {2}}, {{"axis", 0}});
+        const_target_shape_1 = makeConst({batch, head_cnt, seq_len, ndims / 2, 2});
+    } else {
+        ListConstruct_452_Concat =
+            makePattern<opset1::Concat>({seq_length, {-1}, {head_cnt}, {ndims / 2}, {2}}, {{"axis", 0}});
+        const_target_shape_1 = makeConst({seq_len, batch, head_cnt, ndims / 2, 2});
+    }
+
+    std::shared_ptr<ov::Node> ListConstruct_379_Concat = nullptr;
+    std::shared_ptr<ov::Node> const_target_shape_2 = nullptr;
+    if (support_2d_rope) {
+        ListConstruct_379_Concat =
+            makePattern<opset1::Concat>({{-1}, {1}, seq_length, {ndims / 2}, {2}}, {{"axis", 0}});
+        const_target_shape_2 = makeConst({batch, 1, seq_len, ndims / 2, 2});
+    } else {
+        ListConstruct_379_Concat =
+            makePattern<opset1::Concat>({seq_length, {-1}, {1}, {ndims / 2}, {2}}, {{"axis", 0}});
+        const_target_shape_2 = makeConst({seq_len, batch, 1, ndims / 2, 2});
+    }
 
     auto reshape_Reshape_453 = makePattern<opset1::Reshape>(
```

Comment:
in some version of the glm-4 model, the `Rehape_435` could use `{0, head_cnt, 0, ndims / 2, 2}` with `special_zero` instead of `{seq_len, batch, head_cnt, ndims / 2, 2})`
```
auto const_x_shape = makeConst({0, head_cnt, 0, ndims / 2, 2});
auto Rehape_435 = makePattern<opset1::Reshape>(
        {slice_Slice_437 | var_split_1->output(0), const_x_shape}, {{"special_zero", true}});
```

Reply:
I think the current condition can find RoPE pattern in latest glm4 models. 
is it necessary to this additional condition?  
Could you please let me know the model with specific pattern which current condition does not find rope ?
thanks.

Reply:
I understand what you mean. For gpu plugin, it's working well but it fail to match on CPU plugin. I will check it this failure.

Reply:
Update code and verified that succeed to find RoPE pattern for glm4 on CPU plugin. please review it again. thanks.